File: Disjoint Set & Heap/jichon/24542.js
const fs = require("fs");
const input = fs.readFileSync("./input.txt").toString().trim().split("\n");

//disjoint set -> 서로소 집합
// union-find
// find : 집합 찾기 -> 루트 노드까지 끝까지 찾음
const find_parent = (parent, x) => {
  if (parent[x] !== x) {
    parent[x] = find_parent(parent, parent[x]);
  }
  return parent[x];
};

// union : 두 원소가 속한 집합 합치기
const union_parent = (parent, a, b) => {
  a = find_parent(parent, a);
  b = find_parent(parent, b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
};

const solution = () => {
  const [N, M] = input[0].split(" ").map(Number);
  const parent = Array.from({ length: N + 1 }, (_, i) => i);

  // 모든 원소에 대해 union 수행
  for (let i = 1; i <= M; i++) {
    const [a, b] = input[i].split(" ").map(Number);
    union_parent(parent, a, b);
  }

  let rootCount = new Map();
  for (let i = 1; i < N + 1; i++) {
    const root = find_parent(parent, i);
    // root node가 count가지면 그거 가져오고 없으면 0으로 대체, 발견된 거 +1 해서 갱신
    rootCount.set(root, (rootCount.get(root) || 0) + 1);
  }

  let total = 1;
  //위에서 저장한 rootCount의 value를 가져옴, 즉 위에서 갱신한 정점의 개수
  for (const size of rootCount.values()) {
    total = (total * size) % 1000000007;
  }
  console.log(total);
};

solution();

File: DisjointSet/24542.py
import sys
input = sys.stdin.readline

def find(x):
    if x == parent[x]:
        return x
    parent[x] = find(parent[x]) # 경로 압축
    return parent[x]

def union(x, y):
    parentX = find(x)
    parentY = find(y)
    
    if parentX == parentY:
        return
    if size[parentX] > size[parentY]:
        parent[parentY] = parentX
        size[parentX] += size[parentY]
    else:
        parent[parentX] = parentY
        size[parentY] += size[parentX]        
    
n, m = map(int, input().strip().split())

parent = [i for i in range(n+1)] # 자기 자신을 parent node
size = [1] * (n+1)

for _ in range(m):
    x, y = map(int, input().strip().split())
    union(x, y)
    
sum = 1
parentSet = set()
for i in parent:
    if i != 0: 
        parentSet.add(i)
        
for i in parentSet:
    sum *= size[i]
        
print(sum % 1000000007)    